tools/djxl_ng_main.cc

```diff
@@ -425,52 +425,44 @@ int main(int argc, char** argv) {
       return EXIT_FAILURE;
     };
     // TODO(firsching): handle non-reconstruct JPEG
-  } else if (codec == jxl::extras::Codec::kPNM) {
-    JxlDataType datatype = JXL_TYPE_UINT8;
-    uint32_t num_channels = 3;
-    if (std::string(extension) == ".pfm") {
-      datatype = JXL_TYPE_FLOAT;
-    }
-    if (std::string(extension) == ".pgm") {
-      num_channels = 1;
-    }
-
-    JxlPixelFormat format = {num_channels, datatype, JXL_NATIVE_ENDIAN, 0};
+  } else {
+    JxlPixelFormat format;
+    std::unique_ptr<jxl::extras::Encoder> encoder =
+        jxl::extras::Encoder::FromExtension(extension);
     jxl::extras::PackedPixelFile ppf;
-    if (DecompressJxlToPackedPixelFile(compressed, ppf, format, std::move(dec),
-                                       std::move(runner)) != 0) {
+    if (DecompressJxlToPackedPixelFile(
+            compressed, ppf, encoder->AcceptedFormats(), std::move(dec),
+            std::move(runner)) != 0) {
+      fprintf(stderr, "DecompressJxlToPackedPixelFile failed\n");
       return EXIT_FAILURE;
     }
-    if (ppf.info.exponent_bits_per_sample != 0) {
-      if (num_channels == 1 && ppf.info.num_color_channels == 3) {
-        JXL_WARNING("For color images, the filename should end with .ppm.\n");
-      }
-      if (num_channels == 3 && ppf.info.num_color_channels == 1) {
-        JXL_WARNING(
-            "For grayscale images, the filename should end with .pgm.\n");
-      }
-      if (ppf.info.bits_per_sample > 16) {
-        JXL_WARNING("PPM only supports up to 16 bits per sample");
-      }
+    if (!jxl::extras::SelectFormat(encoder->AcceptedFormats(), ppf.info,
+                                   &format)) {
+      fprintf(stderr, "SelectFormat failed\n");
+      return EXIT_FAILURE;
+    }
+    jxl::extras::EncodedImage encoded_image;
+    JXL_RETURN_IF_ERROR(encoder->Encode(ppf, &encoded_image));
+    if (ppf.frames.size() != encoded_image.bitstreams.size()) {
+      fprintf(stderr,
+              "didn't receive expected number of frames: %" PRIuS
+              " (expected) versus %" PRIuS "(received)\n",
+              ppf.frames.size(), encoded_image.bitstreams.size());
+      return EXIT_FAILURE;
     }
```

Comment:
I am not completely sure it’s for this function to check that `Encode` returned the correct number of bitstreams (which could be 1 if the format supports animations). I think I would favour trusting `Encode` to return an appropriate number of bitstreams or an error if one occurs, and replacing the instances of `ppf.frames.size()` below with `encoded_image.bitstreams.size()`. What do you think?